src/Common/NamedCollections/NamedCollectionsFactory.cpp

```diff
@@ -152,18 +155,249 @@ bool NamedCollectionFactory::removeIfExistsUnlocked(
+}
+
+void NamedCollectionFactory::removeFromSQL(const ASTDropNamedCollectionQuery & query)
+{
+    std::lock_guard lock(mutex);
+    loadIfNot(lock);
+
+    if (!exists(query.collection_name, lock))
+    {
+        if (query.if_exists)
+            return;
+
+        throw Exception(
+            ErrorCodes::NAMED_COLLECTION_DOESNT_EXIST,
+            "Cannot remove collection `{}`, because it doesn't exist",
+            query.collection_name);
+    }
+
+    metadata_storage->remove(query.collection_name);
+    remove(query.collection_name, lock);
+}
+
+void NamedCollectionFactory::updateFromSQL(const ASTAlterNamedCollectionQuery & query)
+{
+    std::lock_guard lock(mutex);
+    loadIfNot(lock);
+
+    if (!exists(query.collection_name, lock))
+    {
+        if (query.if_exists)
+            return;
+
+        throw Exception(
+            ErrorCodes::NAMED_COLLECTION_DOESNT_EXIST,
+            "Cannot remove collection `{}`, because it doesn't exist",
+            query.collection_name);
+    }
+
+    metadata_storage->update(query);
+
+    auto collection = getMutable(query.collection_name, lock);
+    auto collection_lock = collection->lock();
+
+    for (const auto & [name, value] : query.changes)
+    {
+        auto it_override = query.overridability.find(name);
+        if (it_override != query.overridability.end())
+            collection->setOrUpdate<String, true>(name, convertFieldToString(value), it_override->second);
+        else
+            collection->setOrUpdate<String, true>(name, convertFieldToString(value), {});
+    }
+
+    for (const auto & key : query.delete_keys)
+        collection->remove<true>(key);
+}
+
+void NamedCollectionFactory::reloadFromSQL()
+{
+    std::lock_guard lock(mutex);
+    if (loadIfNot(lock))
+        return;
+
+    auto collections = metadata_storage->getAll();
+    removeById(NamedCollection::SourceId::SQL, lock);
+    add(std::move(collections), lock);
+}
+
+void NamedCollectionFactory::updateFunc()
+{
+    LOG_TRACE(log, "Named collections background updating thread started");
+
+    while (!shutdown_called.load())
+    {
+        NamedCollectionsMap collections;
+        try
+        {
+            reloadFromSQL();
+        }
+        catch (const Coordination::Exception & e)
+        {
+            if (Coordination::isHardwareError(e.code))
+            {
+                LOG_INFO(log, "Lost ZooKeeper connection, will try to connect again: {}",
+                         DB::getCurrentExceptionMessage(true));
+
+                sleepForSeconds(1);
+            }
+            else
+            {
+                tryLogCurrentException(__PRETTY_FUNCTION__);
+                chassert(false);
+            }
+            continue;
+        }
+        catch (...)
+        {
+            DB::tryLogCurrentException(__PRETTY_FUNCTION__);
+            chassert(false);
+            continue;
+        }
+
+        if (shutdown_called.load())
+            break;
+
+        metadata_storage->waitUpdate();
```

Comment:
I see you are waiting for list watch, but there is still a timeout and we could do useless Keeper requests.
Maybe keep track of the `cversion` for the root node and fetch everything again only if it changed (there is an extra request if the collections actually changed, but we avoid doing unnecessary lists)? Because I don't expect these collections to change often.
WDYT?